File: app/src/main/java/com/pietrantuono/podcasts/addpodcast/model/SearchApiRetrofit.java
package com.pietrantuono.podcasts.addpodcast.model;

import com.pietrantuono.podcasts.addpodcast.model.pojos.Podcast;
import com.pietrantuono.podcasts.addpodcast.model.retrofitconverters.PodcastSearchResultConverterFactory;

import java.util.List;
import java.util.Map;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;


public class SearchApiRetrofit implements SearchApi {
    private static final String ITUNES = "https://itunes.apple.com";
    private final SearchApi api;

    public SearchApiRetrofit() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(ITUNES)
                .addConverterFactory(new PodcastSearchResultConverterFactory(GsonConverterFactory.create()))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();

        api = retrofit.create(SearchApi.class);
    }

    @Override
    public Observable<List<Podcast>> search(@retrofit2.http.QueryMap Map<String, String> query) {
        throw new UnsupportedOperationException("Not implemented");
    }

    @Override
    public Observable<SearchResult> search(String query) {
        QueryMap map = new QueryMap(query);
        return api.search(map).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).map(list->new SearchResult(list, query)).cache();
    }

}

File: app/src/main/java/com/pietrantuono/podcasts/addpodcast/model/SearchApiRetrofit.kt
package com.pietrantuono.podcasts.addpodcast.model


import com.pietrantuono.podcasts.addpodcast.model.retrofitconverters.PodcastSearchResultConverterFactory

import models.pojos.Podcast
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

class SearchApiRetrofit : SearchApi {
    private val api: SearchApi

    init {
        val retrofit = Retrofit.Builder()
                .baseUrl(ITUNES)
                .addConverterFactory(PodcastSearchResultConverterFactory(GsonConverterFactory.create()))
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .build()

        api = retrofit.create(SearchApi::class.java)
    }

    override fun search(@retrofit2.http.QueryMap query: Map<String, String>): Observable<List<Podcast>> {
        throw UnsupportedOperationException("Not implemented")
    }

    override fun search(query: String): Observable<SearchResult> {
        val map = QueryMap(query)
        return api.search(map).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).map { list -> SearchResult(list, query) }.cache()
    }

    companion object {
        private val ITUNES = "https://itunes.apple.com"
    }

}
